openedx/core/djangoapps/catalog/management/commands/cache_programs.py

```diff
@@ -24,18 +30,107 @@ class Command(BaseCommand):
+                client = create_catalog_api_client(user)
+            except User.DoesNotExist:
+                logger.error(
+                    'Failed to create API client. Service user {username} does not exist.'.format(username)
+                )
+                raise
+
+            try:
+                querystring = {
+                    'exclude_utm': 1,
+                    'status': ('active', 'retired'),
+                    'uuids_only': 1,
+                }
+
+                logger.info('Requesting program UUIDs.')
+                uuids = client.programs.get(**querystring)
+            except:  # pylint: disable=bare-except
+                logger.error('Failed to retrieve program UUIDs.')
+                raise
+
+            total = len(uuids)
+            logger.info('Received {total} UUIDs.'.format(total=total))
+
+            programs = {}
+            failure = False
+            for uuid in uuids:
+                try:
+                    logger.info('Requesting details for program {uuid}.'.format(uuid=uuid))
+                    program = client.programs(uuid).get(exclude_utm=1)
+
+                    cache_key = PROGRAM_CACHE_KEY_TPL.format(uuid=uuid)
+                    programs[cache_key] = program
+                except:  # pylint: disable=bare-except
+                    logger.exception('Failed to retrieve details for program {uuid}.'.format(uuid=uuid))
+                    failure = True
+
+                    continue
+
+            successful = len(programs)
+            logger.info('Caching details for {successful} programs.'.format(successful=successful))
+            cache.set_many(programs, None)
+
+            logger.info('Caching UUIDs for {total} programs.'.format(total=total))
+            cache.set(PROGRAM_UUIDS_CACHE_KEY, uuids, None)
 
+            if failure:
+                # This will fail a Jenkins job running this command, letting site
+                # operators know that there was a problem.
+                sys.exit(1)
+
+        else:
+            failure = False
+            logger.info('populate-multitenant-programs switch is ON')
+
+            catalog_integration = CatalogIntegration.current()
+            username = catalog_integration.service_username
+
+            try:
+                user = User.objects.get(username=username)
+            except User.DoesNotExist:
+                logger.error(
+                    'Failed to create API client. Service user {username} does not exist.'.format(username)
+                )
+                raise
+
+            programs = {}
+            for site in Site.objects.all():
+                if not hasattr(site, "configuration") or not site.configuration.get_value("COURSE_CATALOG_API_URL"):
+                    logger.info('Skipping site {sitename}. No configuration.'.format(sitename=site.domain))
+                    continue
+
+                logger.info('trying site {sitename}'.format(sitename=site.domain))
```

Comment:
Make this message clearer. These messages need to be useful to someone who may not have written this code. In this case, this should at least explain what's about to happen: `Caching programs for site {name}.`